Support LeftSemiJoin and LeftAntiJoin for look up join #37240

khalibartan opened the pull request:
@jordanlewis I'm still working on this. Can you verify if tests are correct? Post that I'll send the changes

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37580)
<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=37580) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=37580) before we can accept your contribution.<br/><hr/>**Utkarsh Gupta** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cockroachdb/cockroach?pullRequest=37580) it.</sub>

khalibartan commented on pkg/sql/distsqlrun/joinreader_test.go:
```diff
@@ -268,6 +268,38 @@ func TestJoinReader(t *testing.T) {
+			joinType:    sqlbase.LeftSemiJoin,
+			inputTypes:  []types.T{*types.Int, *types.String},
+			outputTypes: sqlbase.TwoIntCols,
+			expected:    "[[0 'two']]",
```
Should correct output be `[[0 'two'] [0 'two'] [0 'two'] [0 'two'] [0 'two'] [0 'two'] [0 'two'] [0 'two'] [0 'two']]` because we have 9 instances of 0 in the table?

jordanlewis replied:
Yes, if the left side has n tuples, each of which has a match on the right side, there should be n copies of the left side in the output.

khalibartan commented:
@jordanlewis I'm done with the changes can you please review.

jordanlewis reviewed (commented):
@khalibartan, thanks - the changes aren't complete yet, though - there's no logic for semijoin. Semijoin needs to return the left row if there is a match on the right, once per left row. The logic is not doing that at the moment.

You should take a look at the test cases in `pkg/sql/distsqlrun/mergejoiner_test.go` for semi and anti join for more examples. Those tests should work the same for lookup join as they do for merge join.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37580#-:-LfR9pwkALGWDtgrLvRI:b-bhql0w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @khalibartan)



<!-- Sent from Reviewable.io -->

khalibartan commented:
@jordanlewis Please review.

solongordon reviewed (changes requested):
Thank you for the contribution! I noted some opportunities to simplify the implementation and make it more efficient.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37580#-:-Lg4PyZlEHlqeH5JvXAr:b-a1wli4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @khalibartan and @solongordon)

___
*[pkg/sql/distsqlrun/joinreader.go, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37580#-Lg4LpLjAGfYgBfi9q_J:-Lg4LpLjAGfYgBfi9q_K:bkmn8o3) ([raw file](https://github.com/cockroachdb/cockroach/blob/acc147c389528f55e6fdb8d001cd929ab6339e59/pkg/sql/distsqlrun/joinreader.go#L105)):*
> ```Go
> 	finalLookupBatch        bool
> 	toEmit                  sqlbase.EncDatumRows
> 	inputRowIndicesToKey    map[int]string
> ```

Hm, I'm not convinced this data structure is necessary. It seems like the main thing we need to know is whether an input row was joined to any index rows, which we should be able to determine from `inputRowIdxToOutputRows`.
___
*[pkg/sql/distsqlrun/joinreader.go, line 446 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37580#-Lg4MX5_61LtDmHw63Pg:-Lg4MX5_61LtDmHw63Ph:b-5f78ns) ([raw file](https://github.com/cockroachdb/cockroach/blob/acc147c389528f55e6fdb8d001cd929ab6339e59/pkg/sql/distsqlrun/joinreader.go#L446)):*
> ```Go
> 		for _, inputRowIdx := range jr.keyToInputRowIndices[lookupRow.key] {
> 			_, ok := jr.inputRowIndicesToKey[inputRowIdx]
> 			// Only add to output if jointType is  not LeftSemiJoin or if we have not gotten the match
> ```

Nit: Typo in "jointType" and extra space in `is  not`
___
*[pkg/sql/distsqlrun/joinreader.go, line 448 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37580#-Lg4PPBC2ICrFL-H1Flt:-Lg4PPBC2ICrFL-H1Flu:bwq1sjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/acc147c389528f55e6fdb8d001cd929ab6339e59/pkg/sql/distsqlrun/joinreader.go#L448)):*
> ```Go
> 			// Only add to output if jointType is  not LeftSemiJoin or if we have not gotten the match
> 			if !isJoinTypeLeftSemiJoin || ok {
> 				renderedRow, err := jr.render(jr.inputRows[inputRowIdx], lookupRow.row)
> ```

In the case of an anti join, it seems inefficient to render rows which will never be outputted. Could you see if there is a simple way to avoid this? If not, I think it's fine to just add a TODO noting the inefficiency.
___
*[pkg/sql/distsqlrun/joinreader.go, line 450 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37580#-Lg4LIW29Y5T4S_pD5Lq:-Lg4LIW29Y5T4S_pD5Lr:bmbfgd3) ([raw file](https://github.com/cockroachdb/cockroach/blob/acc147c389528f55e6fdb8d001cd929ab6339e59/pkg/sql/distsqlrun/joinreader.go#L450)):*
> ```Go
> 				renderedRow, err := jr.render(jr.inputRows[inputRowIdx], lookupRow.row)
> 				if isJoinTypeLeftSemiJoin {
> 					delete(jr.inputRowIndicesToKey, inputRowIdx)
> ```

I believe you should only delete the key if `renderedRow != nil`. The reason is that there may be an ON condition which has not yet been evaluated. If this condition fails, `render` will return `nil` and we should not consider the row to have matched. Would be great to add a unit test for this scenario.
___
*[pkg/sql/distsqlrun/joinreader.go, line 464 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37580#-Lg4OwYfAP_e6XThYjqX:-Lg4OwYfAP_e6XThYjqY:b-aopahg) ([raw file](https://github.com/cockroachdb/cockroach/blob/acc147c389528f55e6fdb8d001cd929ab6339e59/pkg/sql/distsqlrun/joinreader.go#L464)):*
> ```Go
> 		}
> 	}
> 	// if joinType was LeftSemiJoin update the map to look for indexes we haven't seen so far
> ```

I might be missing something but I don't see why this is necessary. I tried deleting this block and all your tests still passed. Could you check if it's necessary and if so add a test which exercises this?
___
*[pkg/sql/distsqlrun/joinreader.go, line 485 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37580#-Lg4PGHUD3q-O0w2q96a:-Lg4PGHUD3q-O0w2q96b:b-83b0hf) ([raw file](https://github.com/cockroachdb/cockroach/blob/acc147c389528f55e6fdb8d001cd929ab6339e59/pkg/sql/distsqlrun/joinreader.go#L485)):*
> ```Go
> 	for i, outputRows := range jr.inputRowIdxToOutputRows {
> 		if len(outputRows) == 0 {
> 			if (jr.joinType == sqlbase.LeftOuterJoin) || (jr.joinType == sqlbase.LeftAntiJoin) {
> ```

Nit: Unnecessary parens


<!-- Sent from Reviewable.io -->

khalibartan commented:
@solongordon I'm not sure but wouldn't we need to evaluate the ON condition even in case of anti-join?

solongordon commented:
Good point, I didn't think of that. So you can disregard my comment about avoiding render in anti joins.

(In practice I'm actually not certain if these ON conditions get set on semi- and anti-joins, but for now let's assume that they might.)

khalibartan commented on pkg/sql/distsqlrun/joinreader.go:
```diff
@@ -480,7 +500,9 @@ func (jr *joinReader) collectOutputRows() joinReaderState {
-		if jr.finalLookupBatch {
+		// if joinType was LeftSemiJoin, we would have already removed all the matching keys
+		// if no keys are left match skip and read next batch of input
+		if jr.finalLookupBatch || len(jr.keyToInputRowIndices) == 0 {
```
@solongordon I added this condition to avoid reading the right side if we know we have already match all left indexes for the batch. Do you  want me to do this or remove it?

khalibartan committed:
sql/distsqlrun: Support left semi and anti join in lookup join

Ref: #37240

Added support for LeftSemiJoin and LeftAntiJoin for lookup join. Added
tests for the same.

Release note: None

khalibartan commented:
@solongordon @jordanlewis please review the changes

solongordon reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I think it would be great to revive your optimization in a more straightforward manner, but that can be in a separate PR if you'd like.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37580#-:-LgZ3u6Q750YR636gYKS:b-2vctg3)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @khalibartan)

___
*[pkg/sql/distsqlrun/joinreader.go, line 505 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37580#-LfQGq26sxT2pqD4TOVy-r2-505:-LgZ1S1N1PBBmj4hwWUV:b-9j2cdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/acc147c389528f55e6fdb8d001cd929ab6339e59/pkg/sql/distsqlrun/joinreader.go#L505)):*
<details><summary><i>Previously, khalibartan (Utkarsh) wrote…</i></summary><blockquote>

@solongordon I added this condition to avoid reading the right side if we know we have already match all left indexes for the batch. Do you  want me to do this or remove it?
</blockquote></details>

Ah I see. This sounds like a good optimization but is there a more straightforward way to implement it? Perhaps you could add matched indices to a `FastIntSet` and then check if its size matches the size of the input?


<!-- Sent from Reviewable.io -->

khalibartan commented:
@solongordon I will explore that in separate PR. If everything looks good to you then you can merge it.

solongordon commented:
Great! Just pushed a small change to make the linter happy and will merge once the build passes.

solongordon commented:
bors r+
